add matrix end to end encryption support

josephcrowell opened the pull request:
Compilation will break including in CI. To Avoid that, there are 2 options.

- Add the libolm c library and c build tools so mautix.go can use it.
- Use the ```-tags goolm``` flag to use a pure Go reimplementation of libolm, which removes the need for olm-dev and a C compiler.

In my opinion, the goolm option is better. There are a few threads discussing the options here: https://github.com/mautrix/go/issues/262

poVoq commented:
The thread you linked is not so clear on the status of goolm. Is it actively maintained and working?

I think since we have to terminate e2ee on the server anyways, possible security concerns are not really relevant in our case as long as it works.

josephcrowell commented:
I have been using libolm but I just switched to goolm and yes I have a quite active set of channels that it'll be getting used on now that I've switched.

poVoq commented:
Hmm, I see that you are introducing an new sqlite db to store the session. This probably needs some extra thought in container based environments.

Also is there anything else Matterbridge currently stores in a DB? I thought so far that it is stateless, no? Do you think it would be also possible to store it in the config file instead?

josephcrowell commented:
mautrix.go uses and requires the sqlite db internally. It's either that or in memory which gets lost between sessions, which defeats the purpose. It stores some machine identifiers to keep the bot's login session from identifying as a new machine on every login.

It may me preferable to provide a read/write directory to the container in addition to a read only config file.
